master/internal/command/command.go

```diff
@@ -146,357 +139,326 @@ func remakeCommandsByType(
 }
 
-func restoreCommandsByType(
-	ctx *actor.Context,
-	pgDB *db.PgDB,
+func commandFromSnapshot(
+	db *db.PgDB,
 	rm rm.ResourceManager,
-	taskType model.TaskType,
-) error {
-	commands, err := remakeCommandsByType(ctx, pgDB, rm, taskType)
-	if err != nil {
-		return err
-	}
-
-	for _, cmd := range commands {
-		a, ok := ctx.ActorOf(cmd.taskID, cmd)
-		if !ok {
-			return fmt.Errorf("failed to recreate restored generic command actor %s", cmd.taskID)
-		}
-
-		ctx.Ask(a, actor.Ping{}).Get()
-		ctx.Log().Debugf("restored generic command %s", cmd.taskID)
+	snapshot *CommandSnapshot,
+) *command {
+	taskID := snapshot.TaskID
+	taskType := snapshot.Task.TaskType
+	jobID := snapshot.Task.Job.JobID
+	return &command{
+		db:                 db,
+		rm:                 rm,
+		registeredTime:     snapshot.RegisteredTime,
+		GenericCommandSpec: snapshot.GenericCommandSpec,
+		taskID:             taskID,
+		taskType:           taskType,
+		jobType:            snapshot.Task.Job.JobType,
+		jobID:              jobID,
+		restored:           true,
+		logCtx: logger.Context{
+			"job-id":    jobID,
+			"task-id":   taskID,
+			"task-type": taskType,
+		},
+		syslog: logrus.WithFields(logrus.Fields{
+			"component": "command",
+			"taskID":    taskID,
+		}),
 	}
-
-	return nil
 }
 
-func tryRestoreCommandsByType(
-	ctx *actor.Context,
-	pgDB *db.PgDB,
-	rm rm.ResourceManager,
+// createGenericCommand creates NTSC commands and persists them to the database.
+func (cs *commandService) createGenericCommand(
+	ctx context.Context,
+	taskID model.TaskID,
 	taskType model.TaskType,
-) {
-	err := restoreCommandsByType(ctx, pgDB, rm, taskType)
-	if err != nil {
-		ctx.Log().WithError(err).Warnf("failed to restoreCommandsByType: %s", taskType)
-	}
-}
-
-// command is executed in a containerized environment on a Determined cluster.
-type command struct {
-	mu sync.Mutex
-
-	db *db.PgDB
-	rm rm.ResourceManager
-
-	tasks.GenericCommandSpec
+	jobID model.JobID,
+	jobType model.JobType,
+	spec *tasks.GenericCommandSpec,
+	contextDirectory []byte,
+) (*command, error) {
+	spec.TaskType = taskType
+	cmd := &command{
+		db: cs.db,
+		rm: cs.rm,
 
-	registeredTime time.Time
-	taskID         model.TaskID
-	taskType       model.TaskType
-	jobType        model.JobType
-	jobID          model.JobID
-	allocationID   model.AllocationID
-	lastState      task.AllocationState
-	exitStatus     *task.AllocationExited
-	restored       bool
+		GenericCommandSpec: *spec,
 
-	contextDirectory []byte // Don't rely on this being set outsides of PreStart non restore case.
+		taskID:           taskID,
+		taskType:         taskType,
+		jobType:          jobType,
+		jobID:            jobID,
+		contextDirectory: contextDirectory,
+		logCtx: logger.Context{
+			"job-id":    jobID,
+			"task-id":   taskID,
+			"task-type": taskType,
+		},
+		syslog: logrus.WithFields(logrus.Fields{
+			"component": "command",
+			"task-id":   taskID,
+		}),
+	}
+	// Add it to the registry.
+	cs.commands[cmd.taskID] = cmd
```

Comment:
needs locks

Reply:
reworked it, but `createGenericCommand` is a private function, and wherever it's called (the public 'LaunchNotebook/shell/command/tb' functions) locks the command service before calling it.